coveo-functools/README.md

```diff
@@ -4,116 +4,312 @@ Introspection, finalizers, delegates, dispatchers, waiters...
-def typical_function(test: str) -> str:
-    return test
+TL;DR: Given that `payload` is a dict,`flex.deserialize(payload, hint=Job)` will convert `payload` into an instance of `Job`.
 
-assert flex(typical_function)(**PAYLOAD) == "SUCCESS"
+Here's an example puzzle! An uncanny API returns a messy "transaction" JSON:
+
+```json
+{
+    "Sold_To": {"Name": "Jon"},
+    "Items": [
+        {"sku": 123, "price": 19.99},
+        {"sku": 234, "price": 13.99},
+        {"sku": 0, "price": 0.50, "NOTE": "Forgot the reusable bag at home!!"}
+    ],
+    "Rebates": {
+        "airmiles": {"Flat": 10.0},
+        "coupon": [{"Flat": 0.79}, {"Flat":  1.50}],
+        "senior": {"Percentage": 2.5}
+    },
+    "Id": "GgfhAs89876yh.z"
+}
 ```
 
-Let's see a more practical example:
+
+Wouldn't it be convenient if you could create simple classes/dataclasses around them without any boilerplate?
+
+
+There are many  But you can solve it with flex. In one line, too!
+
+Start by designing a hierarchy of classes with annotations that closely follows the API reference.
+Remember, casing and underscore are ignored in flex, so you could use pep8 if you want:
 
 ```python
+# models.py
+
 from dataclasses import dataclass
+from typing import List, Dict, Union, Optional
 
-import requests  # noqa
-from coveo_functools.flex import flex
+
+class SkuItem:
+    def __init__(self, sku: int, price: float) -> None:
+        self.sku = sku
+        self.price = price
+
+
+@dataclass
+class Rebate:
+    percentage: Optional[float] = None
+    flat: Optional[float] = None
 
 
 @dataclass
-class Owner:
-    login: str
+class Customer:
+    name: str
+
+
+@dataclass
+class Transaction:
+    sold_to: Customer
+    items: List[SkuItem]
+    rebates: Dict[str, Union[Rebate, List[Rebate]]]
+```
+
+Did you notice any flex-related boilerplate in the snippet above? No? Good! :)
+
+Here's how you can use flex's deserializer to bend the furious API's response into your perfect python classes:
+
+
+
+```python
+payload = {
+    "Sold_To": {"Name": "Jon"},
+    "Items": [
+        {"sku": 123, "price": 19.99},
+        {"sku": 234, "price": 13.99},
+        {"sku": 0, "price": 0.50, "NOTE": "Forgot the reusable bag at home!!"}
+    ],
+    "Rebates": {
+        "airmiles": {"Flat": 10.0},
+        "coupon": [{"Flat": 0.79}, {"Flat":  1.50}],
+        "senior": {"Percentage": 2.5}
+    },
+    "Id": "GgfhAs89876yh.z"
+}
+
+one_transaction = flex.deserialize(payload, hint=Transaction)
+list_transactions = flex.deserialize([payload, payload], hint=List[Transaction])
+```
 
+Interesting details:
+- Well, the casing worked! :shrug:
+- `Id` and `NOTE` were dropped because they were excluded from the `Transaction` model. Time saver; some APIs return _tons_ of data.
+- The rebates actually kept the keys, and created `Rebate` instances as the values.
+- In some cases, rebates is a `Rebate` instance, sometimes it's a `List[Rebate]`; it follows whatever the API returns because it was annotated as such.

```

Comment:
_rebates_ in plural with `is` and `it's` makes me uncomfortable. :sweat_smile: I'll let you reword that.